src/pick.c

```diff
@@ -648,93 +633,79 @@ restore_tty(void)
 }
 
 void
-put_line(char *string, int length)
+print_line(const char *string, size_t length, int standout,
+    ssize_t underline_start, ssize_t underline_end)
 {
-	int	i;
+	size_t	i;
+	int	c, col, in_esc_seq = 0, non_printable = 0;
+
+	if (standout)
+		tty_putp(enter_standout_mode);
 
-	for (i = 0; i < length && i < columns; ++i) {
-		if (tty_putc(string[i]) == EOF)
+	for (col = i = 0; i < length && col < columns; i++) {
+		if (i == (size_t)underline_start)
+			tty_putp(enter_underline_mode);
+
+		/*
+		 * Count the number of outputted ANSI escape sequences
+		 * in order to adjust the column count since they do not
+		 * occupy any screen real-estate.
+		 */
+		if (in_esc_seq) {
+			non_printable++;
+			if (string[i] >= '@' && string[i] <= '~')
+				in_esc_seq = 0;
+		} else if (i > 0 && string[i - 1] == ESCAPE
+		    && string[i] == '[') {
+			in_esc_seq = 1;
+			non_printable = 2;
+		}
+
+		c = string[i];
+		switch (c) {
```

Comment:
Since there are only two cases, do you think an if statement would make this clearer?

```
if (c == '\0') {
	/*
 	 * A null character will be present prior the
 	 * terminating null character if descriptions is
 	 * enabled.
 	 */
 	c = ' ';
 	col++;
} else if (!isu8cont(c))
	col++;
```